components/builder-api-client/src/lib.rs

```diff
@@ -292,6 +310,31 @@ pub trait BuilderAPIProvider: Sync + Send {
 
     fn transfer_origin_ownership(&self, origin: &str, token: &str, account: &str) -> Result<()>;
 
+    fn accept_origin_invitation(&self, origin: &str, token: &str, invitation_id: i64)
```

Comment:
I chose a signed integer for a couple of reasons:

* That is the upstream [implementation](https://github.com/habitat-sh/builder/blob/87a4bb970dea81b7e20c1f0aae66ebdb1a4240bf/components/builder-db/src/models/invitations.rs#L19) in Builder
* I'd rather have it be a primitive type that doesn't require an allocation from the heap. As an i64 it stays on the stack and doesn't incur any avoidable overhead such as it might if were a String.

Reply:
Happy to convert to u64.

Reply:
:thinking: I suspect that implementation type is due to the underlying database field and/or the ORM library Builder is using, which doesn't _necessarily_ need to dictate how we represent it at the edges of the system. However, the input is being properly validated and errors are returned to the user in as friendly a manner as possible (i.e., no panic-induced stack traces), this is fine. Anytime I see a signed numeric type when there's no apparent reason for it to be signed, it always raises an eyebrow :sweat_smile: 

And now, if you'll pardon a bit of a digression....

When I say a "real type", there are ways that you can create new types in Rust that don't carry any extra overhead. One pattern you will often see is called the "newtype" pattern, after the similar pattern from Haskell. Here's a nice example:

https://github.com/habitat-sh/habitat/blob/a4d55f980cdad74d91d56344a91d069d816399f4/components/core/src/os/process.rs#L34-L73

Under the hood, that's going just going be a `u32`, but one that has a lot of very unique and constrained behavior that actually makes it much easier to deal with this particular information. We actually have a lot of other examples in the codebase, though many are "hiding" under our [env_config! macro](https://github.com/habitat-sh/habitat/blob/b83cfd8b7e6ece6dc1017f10d54cc783c1db534e/components/core/src/env.rs). 

So creating a new type doesn't inherently mean that you're going to be consuming heap memory. This can be a very good way of efficiently modeling a particular domain, and brings a lot of type safety with it. The canonical example of this would be something like creating a `Pound` and `Kilogram` type. Both could basically be floats underneath, but a function that is expecting `Kilogram`s would never be able to take a `Pound` argument (and vice versa), thus saving you from an entire class of nasty errors. Same amount of memory as a bare float, but with lots of added information that your code can take advantage of.

(Also, this is an API client library, and isn't likely to be, say, used in super tight loops where something like repeated string allocation might end up being a true performance issue. Unless we know for sure that something is going to be an issue, I generally favor getting a good domain model in place, which can be refined later. But, as I said earlier, with Rust you can often get both! :smile:)

Reply:
I've significantly refactored in https://github.com/habitat-sh/habitat/pull/7349/commits/19b27a45acc529f5a8481b9f6f62500b6dd9f75f I'm going to mark this resolved as the code now merits a fresh look.